pricegraph/src/orderbook.rs

```diff
@@ -118,22 +121,29 @@ impl Orderbook {
 
     /// Reduces the orderbook by matching all overlapping ring trades.
     pub fn reduce_overlapping_orders(mut self) -> ReducedOrderbook {
-        Subgraphs::new(self.projection.node_indices()).for_each(|token| loop {
-            match bellman_ford::search(&self.projection, token) {
-                Ok((_, predecessors)) => break predecessors,
-                Err(NegativeCycle(predecessors, node)) => {
-                    let path = path::find_cycle(&predecessors, node, None)
-                        .expect("negative cycle not found after being detected");
-
-                    self.fill_path(&path).unwrap_or_else(|| {
-                        panic!(
-                            "failed to fill path along detected negative cycle {}",
-                            format_path(&path),
-                        )
-                    });
-                }
+        let mut remaining_tokens: BTreeSet<_> = self.projection.node_indices().collect();
+
+        while let Some(&token) = remaining_tokens.iter().next() {
+            remaining_tokens.remove(&token);
+
+            let shortest_path_graph = loop {
+                match ShortestPathGraph::new(&self.projection, token) {
+                    Ok(shortest_path_graph) => break shortest_path_graph,
+                    Err(NegativeCycle(cycle)) => {
```

Comment:
Since `NegativeCycle: Deref<Target = [N]>`, can't we just do:
```suggestion
                    Err(cycle) => {
```

Without having to destructure the `NegativeCycle`?